### PR TITLE
Feature/add order details

### PR DESCRIPTION
**Pull request by chr1sgrant**

### CODE REVIEW

**Comment by x-anderson** [comment]
👍

File: antha/AnthaStandardLibrary/Packages/buffers/order/order.go [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+	"github.com/antha-lang/antha/antha/anthalib/wtype"
+	"github.com/antha-lang/antha/antha/anthalib/wunit"
+)
+
+// Option is a type for use as an argument in the SetOrderInfo function.
+type Option string
+
+const (
+	// Force overwriting of Order Details of an LHComponent in the SetOrderInfo function.
+	FORCEUPDATE Option = "FORCE"
```
Comment (ddn0) [inline antha/AnthaStandardLibrary/Packages/buffers/order/order.go]: constants are written LikeThis

File: antha/AnthaStandardLibrary/Packages/buffers/order/order.go [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+)
+
+// Key to look up order details from an LHComponent.
+const ORDERDETAILS = "ORDERDETAILS"
+
+var (
+	errNotFound = fmt.Errorf("No order info found")
```
Comment (ddn0) [inline antha/AnthaStandardLibrary/Packages/buffers/order/order.go]: errors are written like "this" without initial capitalization. In this case, use `errors.New(...)`

File: antha/AnthaStandardLibrary/Packages/buffers/order/order.go [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+	return fmt.Sprintf("%+v", s)
+}
+
+// GetOrderDetails returns order Details for a component.
+func GetOrderDetails(comp *wtype.LHComponent) (orderDetails Details, err error) {
+
+	order, found := comp.Extra[ORDERDETAILS]
+
+	if !found {
+		return orderDetails, errNotFound
```
Comment (ddn0) [inline antha/AnthaStandardLibrary/Packages/buffers/order/order.go]: in the same function, do not mix naked returns and normal ones.

in this case: 
```
  return Details{}, errNotFound
```

File: antha/AnthaStandardLibrary/Packages/buffers/order/order.go [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+
+	bts, err = json.Marshal(order)
+	if err != nil {
+		return
+	}
+
+	err = json.Unmarshal(bts, &orderDetails)
+
+	if err != nil {
+		err = fmt.Errorf("Problem getting %s order details. Error: %s", comp.Name(), err.Error())
```
Comment (ddn0) [inline antha/AnthaStandardLibrary/Packages/buffers/order/order.go]: errors are written like this: "initial lower case detail colon: root cause"

File: antha/AnthaStandardLibrary/Packages/buffers/order/order.go [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+
+	// look for existing order details
+	existingDetails, err := GetOrderDetails(comp)
+
+	if err == errNotFound {
+		comp.Extra[ORDERDETAILS] = orderDetails
+		return comp, nil
+	} else if err == nil {
+		if !inOptions(FORCEUPDATE, options) {
+			return comp, fmt.Errorf("component %s already contains order info %v. Use order FORCEUPDATE to override.", comp.Name(), existingDetails)
```
Comment (ddn0) [inline antha/AnthaStandardLibrary/Packages/buffers/order/order.go]: preference: callers know the arguments they are giving this function, so allow them to choose to print the object if they want that in their error message:
```
  return nil, errors.New("component already contains order details")
```

File: antha/AnthaStandardLibrary/Packages/buffers/order/order.go [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+		}
+	} else {
+		return comp, err
+	}
+
+	comp.Extra[ORDERDETAILS] = orderDetails
```
Comment (ddn0) [inline antha/AnthaStandardLibrary/Packages/buffers/order/order.go]: dead code

File: antha/AnthaStandardLibrary/Packages/buffers/order/order.go [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+// SetOrderDetails adds order details to an LHComponent.
+// An error will be returned if order details are already encountered unless the FORCEUPDATE option is used as an Option argument in the function.
+// In which case any existing order details will be overwritten.
+func SetOrderDetails(comp *wtype.LHComponent, orderDetails Details, options ...Option) (*wtype.LHComponent, error) {
+
+	// look for existing order details
+	existingDetails, err := GetOrderDetails(comp)
+
+	if err == errNotFound {
+		comp.Extra[ORDERDETAILS] = orderDetails
```
Comment (ddn0) [inline antha/AnthaStandardLibrary/Packages/buffers/order/order.go]: prefer to error cases return early and normal cases to fall through:
```
if err != nil && err != errNotFound {
  return nil, err
} else if err == nil && !force {
  return nil, errNew
}
// err == nil & force || err == errNotFound
comp[xxxx] = value

File: antha/AnthaStandardLibrary/Packages/buffers/order/order.go [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+// Option is a type for use as an argument in the SetOrderInfo function.
+type Option string
+
+const (
+	// Force overwriting of Order Details of an LHComponent in the SetOrderInfo function.
+	FORCEUPDATE Option = "FORCE"
+)
+
+// Key to look up order details from an LHComponent.
+const ORDERDETAILS = "ORDERDETAILS"
```
Comment (ddn0) [inline antha/AnthaStandardLibrary/Packages/buffers/order/order.go]: package qualify this name

**Comment by chr1sgrant** [comment]
I've fixed all linting errors associated with this PR but there's another PR coming in which updates the PH package so I'll remove the buffers package from the makefile ignore list in that one.

**Pull request merged**